File: js/net/ber.js
/* Copyright (C) 2013-2014 Christian Fischbach <cf@cflib.de>
 *
 * This file is part of cflib.
 *
 * cflib is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * cflib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with cflib. If not, see <http://www.gnu.org/licenses/>.
 */

define([
	'cflib/util/util'
], function(util) {

	// returns [tag, tagLen]
	// returns -1 if not enough data available
	function decodeBERTag(data, start, len)
	{
		if (len < 1) return [-1, 0];
		var tag = data[start] & 0x1F;	// remove Class and P/C
		var tagLen = 1;
		if (tag == 0x1F) {
			++tagLen;
			if (len < tagLen) return [-1, 0];
			tag <<= 8;
			tag |= data[++start];
			while (tag & 0x80) {
				++tagLen;
				if (len < tagLen) return [-1, 0];
				tag <<= 8;
				tag |= data[++start];
			}
		}
		return [tag, tagLen];
	}

	// returns -1 if not enough data available
	// returns -2 if length is undefined (one byte: 0x80)
	// returns -3 if too big length was found
	function decodeBERLength(data, start, len)
	{
		// Is some data available?
		if (len < 1) return [-1, 0];

		// If 8th bit is not set, length is in this byte.
		var b = data[start];
		if ((b & 0x80) === 0) return [b, 1];

		// 8th bit is set, so lower bits hold the size of the length
		var ls = b & 0x7F;
		if (ls > 8) return [-3, 0];
		if (len <= ls) return [-1, 0];

		// check for undefined length
		if (ls === 0) return [-2, 1];

		// check for too big length (signed qint64 overflow)
		b = data[++start];
		if (ls == 8 && ((b & 0x80) !== 0)) return [-3, 0];

		// calculate length
		var lengthSize = ls + 1;
		var retval = b;
		while (--ls > 0) retval = (retval << 8) | data[++start];
		return [retval, lengthSize];
	}

	// returns [valueLen, tag, tagLen, lengthSize]
	// returns -1 if not enough data available
	// returns -2 if length is undefined (one byte: 0x80)
	// returns -3 if too big length was found
	function decodeTLV(data, start, len)
	{
		var rv = decodeBERTag(data, start, len);
		var tag = rv[0];
		var tagLen = rv[1];
		if (tag < 0) return [-1, 0, 0, 0];
		var rv2 = decodeBERLength(data, start + tagLen, len - tagLen);
		return [rv2[0], tag, tagLen, rv2[1]];
	}

	function createTag(tagNo, constructed)
	{
		if (tagNo < 0x1F) return [(constructed ? 0xE0 : 0xC0) | tagNo];

		var rv = [tagNo & 0x7F];
		tagNo >>>= 7;
		while (tagNo > 0) {
			rv.unshift((tagNo & 0x7F) | 0x80);
			tagNo >>>= 7;
		}
		rv.unshift(constructed ? 0xFF : 0xDF);
		return rv;
	}

	// If length < 0 the undefined length is written.
	function encodeBERLength(length)
	{
		// one byte encoding
		if (length < 0)    return [0x80];
		if (length < 0x80) return [length];

		var rv = [length & 0xFF];
		while (length > 0xFF) {
			length >>>= 8;
			rv.unshift(length & 0xFF);
		}
		rv.unshift(rv.length | 0x80);
		return rv;
	}

	function makeTLV(tagNo, constructed, uint8Array)
	{
		var tag = createTag(tagNo, constructed);
		if (uint8Array === null)                    { tag.push(0x81, 0); return new Uint8Array(tag); }
		if (!uint8Array || uint8Array.length === 0) { tag.push(0      ); return new Uint8Array(tag); }

		var valueLen = uint8Array.length;
		var len = encodeBERLength(valueLen);
		var valueStart = tag.length + len.length;
		var rv = new Uint8Array(valueStart + valueLen);
		rv.set(tag);
		rv.set(len, tag.length);
		rv.set(uint8Array, valueStart);
		return rv;
	}

	function Serializer(disableTagNumbering)
	{
		this.tagNo = disableTagNumbering ? 0 : 1;
		this.data = new Uint8Array(0);
	}
	$.extend(Serializer.prototype, {

		add: function(d) {
			var od = this.data;
			this.data = new Uint8Array(od.length + d.length);
			this.data.set(od);
			this.data.set(d, od.length);
		},

		box: function(tagNo) {
			return makeTLV(tagNo, true, this.data);
		},

		n: function() {
			if (this.tagNo === 0) this.add([0xC0, 0x81, 0x00]);
			else                  this.tagNo++;
			return this;
		},

		z: function() {
			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			tag.push(0);
			this.add(tag);
			return this;
		},

		i: function(val) {
			if (!val) return this.n();

			var neg = val < 0;
			if (neg) val = -val - 1;
			var b = val & 0xFF;
			var bytes = [neg ? ~b : b];
			val = val > 0xFFFFFFFF ? Math.floor(val / 256) : val >>> 8;
			while (val > 0) {
				b = val & 0xFF;
				bytes.unshift(neg ? ~b : b);
				val = val > 0xFFFFFFFF ? Math.floor(val / 256) : val >>> 8;
			}
			if ((b & 0x80) !== 0) {
				if (neg) bytes.unshift(0xFF);
				else     bytes.unshift(0);
			}

			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			tag.push(bytes.length);
			this.add(tag.concat(bytes));
			return this;
		},

		f32: function(val) {
			if (!val) return this.n();

			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			tag.push(4);
			this.add(tag);
			var fa = new Float32Array(1);
			fa[0] = val;
			this.add(new Uint8Array(fa.buffer));
			return this;
		},

		f64: function(val) {
			if (!val) return this.n();

			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			tag.push(8);
			this.add(tag);
			var fa = new Float64Array(1);
			fa[0] = val;
			this.add(new Uint8Array(fa.buffer));
			return this;
		},

		s: function(str) {
			if (str == '') return this.z();
			if (!str)      return this.n();

			var u8 = util.toUTF8(str);
			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			this.add(tag.concat(encodeBERLength(u8.length)));
			this.add(u8);
			return this;
		},

		a: function(byteArray, emptyIsNull) {
			if (!byteArray            ) return this.n();
			if (byteArray.length === 0) return emptyIsNull ? this.n() : this.z();

			var tag = createTag(this.tagNo === 0 ? 0 : this.tagNo++);
			this.add(tag.concat(encodeBERLength(byteArray.length)));
			this.add(byteArray);
			return this;
		},

		o: function(obj) {
			if (!obj) return this.n();
			return this.a(obj.__serialize(), true);
		},

		map: function(list, func) {
			if (!list) return this.n();
			var S = new Serializer(true);
			for (var i = 0, l = list.length ; i < l ; ++i) func(list[i], S);
			return this.a(S.data, true);
		}

	});

	function Deserializer(data, disableTagNumbering)
	{
		this.data = data;
		this.start = 0;
		this.len = data ? data.length : 0;
		this.tagNo = disableTagNumbering ? 0 : 1;
	}

	$.extend(Deserializer.prototype, {

		read: function() {
			for (;;) {
				// read tlv
				var tlv = decodeTLV(this.data, this.start, this.len);
				var valueLen = tlv[0];
				var tag = tlv[1];
				var tagLen = tlv[2];
				var lengthSize = tlv[3];
				var tlvLen = tagLen + lengthSize + valueLen;
				if (valueLen < 0 || this.len < tlvLen) break;

				// check tag and deserialize
				if (tag == this.tagNo) {
					var rv = valueLen === 0 && lengthSize == 2 ? null :
						new Uint8Array(this.data.buffer, this.data.byteOffset + this.start + tagLen + lengthSize, valueLen);
					this.start += tlvLen;
					this.len   -= tlvLen;
					if (this.tagNo > 0) ++this.tagNo;
					return rv;
				} else if (tag > this.tagNo) {
					if (this.tagNo > 0) ++this.tagNo;
					return null;
				}

				// tag < this.tagNo
				this.start += tlvLen;
				this.len   -= tlvLen;
			}

			// some error occured
			this.len = 0;
			return null;
		},

		i: function() {
			var raw = this.read();
			if (!raw) return 0;
			var s = 0;
			var e = raw.length;
			var rv = raw[s++];
			var neg = (rv & 0x80) !== 0;
			if (neg) rv = ~rv & 0xFF;
			while (s < e) rv = (rv * 256) + (neg ? ~raw[s++] & 0xFF : raw[s++]);
			return neg ? -rv - 1 : rv;
		},

		f32: function() {
			var raw = this.read();
			if (!raw || raw.length != 4) return 0;
			var a = new Uint8Array(4);
			a.set(raw);
			return (new Float32Array(a.buffer))[0];
		},

		f64: function() {
			var raw = this.read();
			if (!raw || raw.length != 8) return 0;
			var a = new Uint8Array(8);
			a.set(raw);
			return (new Float64Array(a.buffer))[0];
		},

		s: function() {
			var raw = this.read();
			if (!raw) return null;
			if (raw.length === 0) return '';
			return util.fromUTF8(raw);
		},

		a: function() {
			var raw = this.read();
			if (!raw) return null;
			return raw;
		},

		map: function(func) {
			var raw = this.read();
			if (!raw) return [];
			var D = new Deserializer(raw, true);
			var rv = [];
			while (D.len > 0) rv.push(func(D));
			return rv;
		}

	});

	// ========================================================================

	var BER = function() {};
	var ber = new BER();

	ber.decodeTLV = decodeTLV;
	ber.makeTLV = makeTLV;
	ber.S = function(disableTagNumbering) { return new Serializer(disableTagNumbering); };
	ber.D = function(data, disableTagNumbering) { return new Deserializer(data, disableTagNumbering); };

	return ber;
});

File: py/cflib/net/ber.py
# Copyright (C) 2013-2026 Christian Fischbach <cf@cflib.de>
#
# This file is part of cflib.
#
# Licensed under the MIT License.

"""Hand-written BER (ASN.1) serialization for cflib.

This is a direct port of ``js/net/ber.js`` and produces byte-identical output to
the JavaScript and C++ implementations.  It deliberately avoids any third party
encoding library; only Python built-ins (``bytearray``, ``struct``, the UTF-8
codec) are used.

Wire model: every value is a TLV (tag / length / value).  Inside a structure the
serializer auto-numbers tags starting at 1 (``tag_no``); ``tag_no == 0`` disables
numbering so every element shares tag 0 (used for list/map elements).
"""

import struct

__all__ = ["Serializer", "Deserializer", "S", "D", "decode_tlv", "make_tlv"]


# ---------------------------------------------------------------------------
# low level tag / length helpers
# ---------------------------------------------------------------------------

def create_tag(tag_no, constructed=False):
    """Encode a tag number (private class) as bytes."""
    if tag_no < 0x1F:
        return bytes([(0xE0 if constructed else 0xC0) | tag_no])

    rv = [tag_no & 0x7F]
    tag_no >>= 7
    while tag_no > 0:
        rv.insert(0, (tag_no & 0x7F) | 0x80)
        tag_no >>= 7
    rv.insert(0, 0xFF if constructed else 0xDF)
    return bytes(rv)


def encode_ber_length(length):
    """Encode a definite length.  A negative length yields the indefinite form."""
    if length < 0:
        return b"\x80"
    if length < 0x80:
        return bytes([length])

    rv = [length & 0xFF]
    while length > 0xFF:
        length >>= 8
        rv.insert(0, length & 0xFF)
    rv.insert(0, len(rv) | 0x80)
    return bytes(rv)


def make_tlv(tag_no, constructed=False, value=None):
    """Build a complete TLV frame.

    ``value`` of ``None`` encodes an explicit null (length ``0x81 00``); an empty
    bytes object encodes an empty value (length ``00``).
    """
    tag = create_tag(tag_no, constructed)
    if value is None:
        return tag + b"\x81\x00"
    if len(value) == 0:
        return tag + b"\x00"
    return tag + encode_ber_length(len(value)) + bytes(value)


def decode_ber_tag(data, start, length):
    """Return ``(tag, tag_len)``; ``tag`` is -1 when not enough data is available."""
    if length < 1:
        return -1, 0
    tag = data[start] & 0x1F  # strip class and primitive/constructed bits
    tag_len = 1
    if tag == 0x1F:
        idx = start
        tag_len += 1
        if length < tag_len:
            return -1, 0
        idx += 1
        tag = (tag << 8) | data[idx]
        while tag & 0x80:
            tag_len += 1
            if length < tag_len:
                return -1, 0
            idx += 1
            tag = (tag << 8) | data[idx]
    return tag, tag_len


def decode_ber_length(data, start, length):
    """Return ``(value_len, length_size)``.

    ``value_len`` is -1 (not enough data), -2 (indefinite length) or -3 (too big).
    """
    if length < 1:
        return -1, 0

    b = data[start]
    if (b & 0x80) == 0:
        return b, 1

    ls = b & 0x7F
    if ls > 8:
        return -3, 0
    if length <= ls:
        return -1, 0
    if ls == 0:
        return -2, 1

    idx = start + 1
    b = data[idx]
    if ls == 8 and (b & 0x80) != 0:
        return -3, 0

    length_size = ls + 1
    retval = b
    while ls > 1:
        ls -= 1
        idx += 1
        retval = (retval << 8) | data[idx]
    return retval, length_size


def decode_tlv(data, start, length):
    """Return ``(value_len, tag, tag_len, length_size)`` for the TLV at ``start``."""
    tag, tag_len = decode_ber_tag(data, start, length)
    if tag < 0:
        return -1, 0, 0, 0
    value_len, length_size = decode_ber_length(data, start + tag_len, length - tag_len)
    return value_len, tag, tag_len, length_size


def _int_to_minimal_bytes(val):
    """Minimal-length big-endian two's complement encoding of a non-zero int."""
    length = 1
    while True:
        try:
            return val.to_bytes(length, "big", signed=True)
        except OverflowError:
            length += 1


# ---------------------------------------------------------------------------
# Serializer
# ---------------------------------------------------------------------------

class Serializer:
    """Builds a BER byte stream.  Chainable: every method returns ``self``."""

    __slots__ = ("tag_no", "_buf")

    def __init__(self, disable_tag_numbering=False):
        self.tag_no = 0 if disable_tag_numbering else 1
        self._buf = bytearray()

    def _tag(self):
        """Return the current tag number, advancing the counter when numbering is on."""
        if self.tag_no == 0:
            return 0
        t = self.tag_no
        self.tag_no += 1
        return t

    def data(self):
        """The serialized payload built so far."""
        return bytes(self._buf)

    def box(self, tag_no):
        """Wrap the whole payload in a constructed TLV with the given tag."""
        return make_tlv(tag_no, True, self._buf)

    def n(self):
        """Null / absent value (skips a tag when numbering, else writes explicit null)."""
        if self.tag_no == 0:
            self._buf += b"\xC0\x81\x00"
        else:
            self.tag_no += 1
        return self

    def z(self):
        """Empty (zero length) value."""
        self._buf += create_tag(self._tag())
        self._buf += b"\x00"
        return self

    def i(self, val):
        if not val:
            return self.n()
        payload = _int_to_minimal_bytes(int(val))
        self._buf += create_tag(self._tag())
        self._buf += bytes([len(payload)])
        self._buf += payload
        return self

    def f32(self, val):
        if not val:
            return self.n()
        self._buf += create_tag(self._tag())
        self._buf += b"\x04"
        self._buf += struct.pack("<f", val)
        return self

    def f64(self, val):
        if not val:
            return self.n()
        self._buf += create_tag(self._tag())
        self._buf += b"\x08"
        self._buf += struct.pack("<d", val)
        return self

    def s(self, val):
        if val is None:
            return self.n()
        if val == "":
            return self.z()
        data = val.encode("utf-8")
        self._buf += create_tag(self._tag())
        self._buf += encode_ber_length(len(data))
        self._buf += data
        return self

    def a(self, val):
        if val is None:
            return self.n()
        if len(val) == 0:
            return self.z()
        self._buf += create_tag(self._tag())
        self._buf += encode_ber_length(len(val))
        self._buf += bytes(val)
        return self

    def _wrap(self, payload):
        """Write ``payload`` as a primitive-class TLV using the current tag."""
        if len(payload) == 0:
            return self.n()
        self._buf += create_tag(self._tag())
        self._buf += encode_ber_length(len(payload))
        self._buf += payload
        return self

    def o(self, obj, ser_func=None):
        """Serialize a nested object (tag numbering restarts at 1 inside)."""
        if obj is None:
            return self.n()
        child = Serializer()
        if ser_func is not None:
            ser_func(obj, child)
        else:
            obj._serialize(child)
        return self._wrap(child._buf)

    def map(self, items, func):
        """Serialize a sequence; elements share tag 0 (numbering disabled)."""
        if not items:
            return self.n()
        child = Serializer(disable_tag_numbering=True)
        for e in items:
            func(e, child)
        return self._wrap(child._buf)

    def p(self, pair, func):
        """Serialize a pair (tag numbering restarts at 1 inside)."""
        if pair is None:
            return self.n()
        child = Serializer()
        func(pair, child)
        return self._wrap(child._buf)


# ---------------------------------------------------------------------------
# Deserializer
# ---------------------------------------------------------------------------

class Deserializer:
    """Reads a BER byte stream produced by :class:`Serializer`."""

    __slots__ = ("data", "start", "len", "tag_no")

    def __init__(self, data, disable_tag_numbering=False):
        self.data = data if data is not None else b""
        self.start = 0
        self.len = len(self.data)
        self.tag_no = 0 if disable_tag_numbering else 1

    def read(self):
        """Read the value bytes for the next expected tag (``None`` if absent)."""
        while True:
            value_len, tag, tag_len, length_size = decode_tlv(self.data, self.start, self.len)
            tlv_len = tag_len + length_size + value_len
            if value_len < 0 or self.len < tlv_len:
                break

            if tag == self.tag_no:
                if value_len == 0 and length_size == 2:
                    rv = None
                else:
                    vstart = self.start + tag_len + length_size
                    rv = self.data[vstart:vstart + value_len]
                self.start += tlv_len
                self.len -= tlv_len
                if self.tag_no > 0:
                    self.tag_no += 1
                return rv

            if tag > self.tag_no:
                if self.tag_no > 0:
                    self.tag_no += 1
                return None

            # tag < self.tag_no -> skip this element and keep looking
            self.start += tlv_len
            self.len -= tlv_len

        self.len = 0
        return None

    def n(self):
        """Skip one tag position."""
        self.tag_no += 1

    def i(self):
        raw = self.read()
        if not raw:
            return 0
        return int.from_bytes(raw, "big", signed=True)

    def f32(self):
        raw = self.read()
        if not raw or len(raw) != 4:
            return 0.0
        return struct.unpack("<f", raw)[0]

    def f64(self):
        raw = self.read()
        if not raw or len(raw) != 8:
            return 0.0
        return struct.unpack("<d", raw)[0]

    def s(self):
        raw = self.read()
        if raw is None:
            return None
        if len(raw) == 0:
            return ""
        return raw.decode("utf-8")

    def a(self):
        raw = self.read()
        if raw is None:
            return None
        return bytes(raw)

    def map(self, func):
        """Deserialize a sequence, calling ``func(D)`` for each element."""
        raw = self.read()
        if not raw:
            return []
        d = Deserializer(raw, disable_tag_numbering=True)
        rv = []
        while d.len > 0:
            rv.append(func(d))
        return rv


# ---------------------------------------------------------------------------
# module level convenience (mirrors the `ber` default export in ber.js)
# ---------------------------------------------------------------------------

def S(disable_tag_numbering=False):
    return Serializer(disable_tag_numbering)


def D(data, disable_tag_numbering=False):
    return Deserializer(data, disable_tag_numbering)
